Snowpark integration: Show an uncollected dataframe when using st.dataframe

sfc-gh-tszerszen opened the pull request:
When using **unevaluated** `snowflake.snowpark.dataframe.DataFrame` Streamlit raises an exception.

This PR makes Streamlit take 10k rows from unevaluated snowpark DataFrame and shows an caption warning to user if we reach 10k rows.

<!--
Before contributing (PLEASE READ!)

⚠️ If your contribution is more than a few lines of code, then prior to starting to code on it please post in the issue saying you want to volunteer, then wait for a positive response. And if there is no issue for it yet, create it first.

This helps make sure:

  1. Two people aren't working on the same thing
  2. This is something Streamlit's maintainers believe should be implemented/fixed
  3. Any API, UI, or deeper architectural changes that need to be implemented have been fully thought through by Streamlit's maintainers
  4. Your time is well spent!

More information in our wiki: https://github.com/streamlit/streamlit/wiki/Contributing
-->

## 📚 Context

_Please describe the project or issue background here_

- What kind of change does this PR introduce?

  - [ ] Bugfix
  - [x] Feature
  - [ ] Refactoring
  - [ ] Other, please describe:

## 🧠 Description of Changes

- _Add bullet points summarizing your changes here_

  - [ ] This is a breaking API change
  - [ ] This is a visible (user-facing) change

**Revised:**

_Insert screenshot of your updated UI/code here_

**Current:**

_Insert screenshot of existing UI/code here_

## 🧪 Testing Done

- [ ] Screenshots included
- [ ] Added/Updated unit tests
- [ ] Added/Updated e2e tests

## 🌐 References

_Does this depend on other work, documents, or tickets?_

- **Issue**: Closes #XXXX

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

lukasmasuch commented on lib/streamlit/elements/arrow.py:
```diff
@@ -73,6 +75,15 @@ def _arrow_dataframe(
             This takes precedence over the width argument.
             This argument can only be supplied by keyword.
 
+        num_rows : int or None
```
I think we don't need to have this as a parameter here since it is anyways just used internally. We usually are very cautious about adding new arguments to our commands.  But maybe this is planned to be supported via the API (@jrieke)?

sfc-gh-tszerszen replied:
They are removed and `st.caption` is used directly instead 👍

lukasmasuch commented on lib/streamlit/elements/dataframe_selector.py:
```diff
@@ -101,9 +102,35 @@ def dataframe(
            height: 410px
 
         """
+        num_rows = None
```
I think this code can be moved into `_arrow_dataframe`. `_legacy_dataframe` isn't supposed to support any new features, and might be removed soon anyways.

sfc-gh-tszerszen replied:
Sure 👍 it's moved 👍

lukasmasuch commented on lib/streamlit/config.py:
```diff
@@ -454,6 +455,17 @@ def _logger_message_format() -> str:
 )
 
+_create_option(
+    "client.snowparkMaxRows",
```
This probably needs to be named slightly more generic so that it can be eventually used by other unevaluated dataframe implementations (e.g. Spark Dataframe) (fyi @jrieke). Also, in general, we try to don't include too many snowflake-specific aspects in the Streamlit core codebase.

sfc-gh-tszerszen replied:
As of a new spec this is outdated 👍

lukasmasuch commented on frontend/src/components/widgets/DataFrame/DataFrame.tsx:
```diff
@@ -463,107 +465,118 @@ function DataFrame({
-            })
+    <div>
+      {showMaxRowsWarning ? (
+        <StreamlitMarkdown
```
I'm actually not sure if we should integrate this directly in the front end. The alternative would be to just execute an `st.caption` directly from the backend with the given text. This would allow us 1) to leave out the `millify` dependency 2) makes it much easier to also support snowpark dataframes for other components that use dataframes as input (e.g. built-in charts) 3) Would not need the (currently) mostly snowflake-specific properties (`num_rows`, `max_rows`) in the arrow proto.

We basically can just check if it is a snowflake dataframe, check the number of rows, and execute an `st.caption` in case `num_rows` > `max_rows`. Eventually, we can move this into utilities (or even into the `convert_anything_to_df` function) to reuse this in a few other components.

sfc-gh-tszerszen replied:
I can check it out, however this will introduce `millify` dependency in our backend, is that ok?

sfc-gh-tszerszen replied:
I think it's better to introduce more dependencies on frontend side and avoid them in the backend, also do calculations and presentation stuff on UI side when possible. @jrieke @kmcgrady do you have an opinion on that?

lukasmasuch replied:
I think we can avoid `millify` in the backend by using our own utility method, e.g.: https://stackoverflow.com/a/3155023 The feature we need from `millify` is probably just a couple lines of code. 

> also do calculations and presentation stuff on UI side when possible

In agree in general. But in this specific case, it adds quite a bit of logic just to render the same message that we can also directly trigger from the backend. If we move this capability into a shared method - e.g. `convert_anything_to_df` - and reuse it across different commands it is even a lot more complicated compared to just calling `st.caption`.

lukasmasuch replied:
Here is another utils method from an old project that might work here:

```python
def simplify_number(num: int) -> str:
    num_converted = float("{:.2g}".format(num))
    magnitude = 0
    while abs(num_converted) >= 1000:
        magnitude += 1
        num_converted /= 1000.0
    return "{}{}".format(
        "{:f}".format(num_converted).rstrip("0").rstrip("."),
        ["", "K", "M", "B", "T"][magnitude],
    )
```

jrieke replied:
I don't really care where this lives but agree with Lukas point on "2) makes it much easier to also support snowpark dataframes for other components that use dataframes as input (e.g. built-in charts)". That's [here](https://www.notion.so/streamlit/Product-spec-045bb2e7008847c08b9308a9329aab9d#e943f56852a348ba847a3300ae8b5e50) in the spec (and we should also make it work for related things like `st.bar_chart` and `st.area_chart`), so would be great to find a universal solution for this problem. (But don't have strong opinions.)

jrieke commented:
@sfc-gh-tszerszen I simplified the spec a bit, which should resolve most annoying questions. See [here](https://www.notion.so/streamlit/Product-spec-045bb2e7008847c08b9308a9329aab9d#de7c8205f6044f71a702efafc563b99c) but also going to paste it below:

- Do not add a parameter `max_rows` or a config option to specify the number of shown rows. We just show 10k rows and that’s it. If devs want to show more rows, they can simply call `collect()` or `take()` on the dataframe themselves. If people complain during SiS PrPr, we can still follow up.
- Do not show the number of actual dataframe rows in the warning (i.e. the “has 144B rows” above). This saves us the call to `count()` (which is annoying for the user since it’s an additional SQL query) and it saves us the dependency on millify.
- Change the warning message to “Showing only 10k rows. Call `collect()` on the dataframe to show more.” Please make sure that this message fits into one line with the default width of Streamlit. If it doesn’t, reach out to Johannes to shorten it. Note that we should not mention anything Snowpark-specific so that we can reuse the same message for PySpark.
- One thing to note: Because we are not querying the length of the original dataframe any more, we can just see if the capped dataframe (coming from `take()`) is 10k rows or less. We should show the warning if the capped length is 10k rows. But this also means if the original, uncapped dataframe is *exactly* 10k rows, we’ll show the warning even though nothing was capped. That’s a tiny edge case though and not super critical, so should be fine.

tconkling reviewed (dismissed):
I think the purpose behind this PR is a good and I haven't actually read the code, so this is not really a "Request for changes" but more a "let's discuss this as a team before going forward":

We're hoping that the open source project can remain as general as possible, and to that end it'd be great to have few (if any) Snowpark-related special cases. 

There may be other ways to achieve this sort of functionality (e.g. a custom component; an abstract "dataframe input handler" interface; etc) that don't involve Snowpark-specific changes to the library. 

And to be clear: this particular change _may_ warrant a special case, I'm not sure. But we should be clear about why we're breaking this convention before doing so.

(@jrieke tagging you here for product discussion!)

jrieke commented:
@tconkling Yes agree to keep the Snowpark specifics as small as possible. But I think this a) is an important thing we should do and b) more importantly, this is not a super-specific Snowpark feature but one that also applies to PySpark and that we probably also want to extend to other lazily evaluated dataframe libraries in the future. 

See my comment above, which will take out most of the Snowpark-specifics in terms of what we expose to the user. There will obviously be a bit of Snowpark specific things we need to do unter the hood, just to make sure this is actually a Snowpark dataframe and we are dealing with it in the right way. Unfortunately, we can't just check if e.g. the dataframe has a `take()` method, since pandas also has that method but it functions differently. (If there is a generic way to do all of this without mentioning anything about Snowpark, that's of course great, but from what I've seen so far, I think that's not easy or possible).

lukasmasuch commented:
If we want to keep the Snowflake-specific logic as minimal as possible, I would argue to just add the Snowflake DataFrame check & collection directly to [convert_anything_to_df](https://github.com/streamlit/streamlit/blob/6d1d6c1ef3130519decbff535479ef3c80f1b889/lib/streamlit/type_util.py#L414). Here is a bit of untested & unfinished pseudo-code (it probably needs a bit more of the logic from the PR):

```python
MAX_SF_DF_ROWS = 10000
if type_util.is_snowpark_dataframe(data):
    collected_df = df.take(MAX_SF_DF_ROWS).collect()
    if collected_df == MAX_SF_DF_ROWS:
        st.caption(f"Showing only {MAX_SF_DF_ROWS} rows. Call `collect()` on the dataframe to show more.")
    
    return collected_df
```

but I think it is probably everything that is needed to support Snowflake DataFrames across all our components and we can do the same for Spark dataframes, etc... This isn't a pretty or optimal long-term solution, but I think it would be a good short-term solution without getting snowflake stuff too much distributed across the codebase. In the long term, I think we anyways need something that can actually handle data streaming or pagination for these kinds of dataframes. And for these capabilities, defining a good dataframe handler abstraction would be very useful.

sfc-gh-tszerszen commented:
@LukasMasuch @jrieke @tconkling thank you for taking a look at this 🙇 

This PR is updated with the spec provided by @jrieke, including:

* @LukasMasuch wrote: 
> I'm actually not sure if we should integrate this directly in the front end. The alternative would be to just execute an st.caption directly from the backend with the given text.

Now `st.caption` is called directly in `st.dataframe` 👍 

> If we want to keep the Snowflake-specific logic as minimal as possible, I would argue to just add the Snowflake DataFrame check & collection directly to [convert_anything_to_df](https://github.com/streamlit/streamlit/blob/6d1d6c1ef3130519decbff535479ef3c80f1b889/lib/streamlit/type_util.py#L414).

Now the logic is moved to [convert_anything_to_df](https://github.com/streamlit/streamlit/blob/716f9478952f753f59ddfe13548e2e0948daeaa7/lib/streamlit/type_util.py) 👍 

@jrieke wrote:

> Do not add a parameter max_rows or a config option to specify the number of shown rows.

Updated now 👍 

> Do not show the number of actual dataframe rows in the warning (i.e. the “has 144B rows” above). This saves us the call to count()

Ok, it's updated now 👍 

> Change the warning message to “Showing only 10k rows. Call collect() on the dataframe to show more.” Please make sure that this message fits into one line with the default width of Streamlit. 

It's really short it first I've added E2E tests for this and everything is fine 👍 

> One thing to note: Because we are not querying the length of the original dataframe any more, we can just see if the capped dataframe (coming from take()) is 10k rows or less. We should show the warning if the capped length is 10k rows. But this also means if the original, uncapped dataframe is exactly 10k rows, we’ll show the warning even though nothing was capped. That’s a tiny edge case though and not super critical, so should be fine.

Good call out, I'm just checking for

```python
            # when needed, notify the user that data was capped
            if data.shape[0] == type_util.MAX_SF_DF_ROWS:
                st.caption(
                    "Showing only 10k rows. Call `collect()` on the dataframe to show more."
                )
```

If that's fine it's done 👍 

Ok, the PR is updated, can I ask @LukasMasuch @jrieke @tconkling for re-review?

sfc-gh-tszerszen commented:
**Update**

I've improved this PR by moving capped detection function into `def show_warning_if_data_is_capped` and using it in the following components:
* st.dataframe
* st.table
* st.line_chart
* st.bar_chart
* st.area_chart

It does not cover all of the components yet, however it's better in my opinion than just showing a warning on `st.dataframe`. What's your opinion on this @jrieke @LukasMasuch ?

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -440,6 +443,18 @@ def convert_anything_to_df(df: Any) -> DataFrame:
+            )
+
+            df = cast(MockSnowparkDataFrame, df)
+        return pd.DataFrame(df.take(MAX_SF_DF_ROWS))
```
nit: did you do any research if `pd.DataFrame(df.take(MAX_SF_DF_ROWS))` is actually the best method for getting a limited dataframe from Snowflake? There is also the `to_pandas` method that can be called on Snowpark data frames, so maybe something like `df.limit(MAX_SF_DF_ROWS).to_pandas()` might be a better option? I'm mentioning this since there might be quite a few issues regarding type detection in the pandas DataFrame if you just insert it via `take`. I have no clue what `to_pandas` does exactly in the background, but maybe it is optimized for converting Snowflake dataframes to Pandas dataframes. If we are not sure about this, we might just reach out to the Snowpark team to ask what they would recommend.

sfc-gh-tszerszen replied:
@LukasMasuch I did investigation and decided to go with `return pd.DataFrame(df.take(MAX_SF_DF_ROWS))` on purpose, if you run the code below (one needs to provide it's snowflake connection params), you will get seg fault :/

```import dotenv

from snowflake.snowpark import Session


def create_session():
    config = dotenv.dotenv_values(".env")
    dotenv.load_dotenv()
    connection_parameters = {
        "account": config["SNOWFLAKE_ACCOUNT"],
        "user": config["SNOWFLAKE_USER"],
        "password": config["SNOWFLAKE_PASSWORD"],
        "role": config["SNOWFLAKE_USER_ROLE"],
        "warehouse": config["SNOWFLAKE_WAREHOUSE"],
        "database": config["SNOWFLAKE_DATABASE"],
        "schema": config["SNOWFLAKE_SCHEMA"],
    }
    return Session.builder.configs(connection_parameters).create()


session = create_session()
df = session.table("SNOWFLAKE_SAMPLE_DATA.TPCDS_SF100TCL.CATALOG_PAGE").limit(100)
print(df.limit(10000).to_pandas())
```

The `return pd.DataFrame(df.take(MAX_SF_DF_ROWS))`  works perfectly fine 👍

lukasmasuch replied:
Ok, sounds good 👍 Thought, it might still be good to have someone from the Snowpark team have a quick look at this, maybe there is something else we are not taking care of.

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -440,6 +443,18 @@ def convert_anything_to_df(df: Any) -> DataFrame:
+    ):
+        if TYPE_CHECKING:
+            # df needs to be cased to MockSnowparkDataFrame when checking for types, because take method is required
+            from tests.streamlit.snowpark_mocks import (
```
I'm not sure if we should import from `tests` in our main codebase - even if it is under a `TYPE_CHECKING` flag. Is there any precedent for doing this? Probably something to discuss with the team. In this case, it is probably better to just ignore the types here.

lukasmasuch commented on lib/streamlit/elements/arrow.py:
```diff
@@ -54,7 +55,7 @@ def _arrow_dataframe(
         Parameters
         ----------
-        data : pandas.DataFrame, pandas.Styler, pyarrow.Table, numpy.ndarray, Iterable, dict, or None
+        data : pandas.DataFrame, pandas.Styler, pyarrow.Table, numpy.ndarray, snowflake.snowpark.DataFrame, Iterable, dict, or None
```
There is also the type definition above which we might want to add `snowflake.snowpark.DataFrame` as well, otherwise people will get typing warnings.

https://github.com/streamlit/streamlit/blob/71eb38ccec573c27811a18cd0b260d1a66bbffa9/lib/streamlit/elements/arrow.py#L41

lukasmasuch replied:
Also, the docstrings and types needs to be added to the dataframe selector methods as well.

sfc-gh-tszerszen replied:
> There is also the type definition above which we might want to add snowflake.snowpark.DataFrame as well, otherwise people will get typing warnings.

I didn't add it to:
```Data = Union[DataFrame, Styler, pa.Table, ndarray, Iterable, Dict[str, List[Any]], None]```
because:
* Snowpark works only in Python 3.8 (we have Min Python Version tests and Max Python versions tests)
* We don't have Snowpark as Streamlit dependency and I guess we don't want to have it as a dep

If people will get warnings I guess we have to add `Any` to the line:
```Data = Union[DataFrame, Styler, pa.Table, ndarray, Iterable, Dict[str, List[Any]], Any, None]```

🤔 What can we do from this point?

lukasmasuch replied:
ahh I see 🤔 You can also add it as a string to the types which doesn't require any dependency:

```python
Data = Union[
    DataFrame,
    Styler,
    pa.Table,
    ndarray,
    Iterable,
    Dict[str, List[Any]],
    "snowflake.snowpark.DataFrame",
    None,
]
```

But I'm not 100% sure if that works, needs to be tested.

sfc-gh-tszerszen replied:
When we add `"snowflake.snowpark.DataFrame",`
`mypy` returns ```lib/streamlit/elements/arrow.py:48: error: Name "snowflake" is not defined  [name-defined]``` in that case 😢 I'm reverting back to `Any` I guess it's better than warnings for people/new dependency for Streamlit (And only for Python 3.8 😱 )

lukasmasuch replied:
alright, lets go with leaving that out for now

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -49,6 +49,9 @@
     from pydeck import Deck
 
+
+MAX_SF_DF_ROWS = 10000
```
nit: I know I suggested this name, but I think we can go a bit more generic here so that we can reuse the same constant for other unevaluated / out-of-memory / out-of-core dataframe implementations. Just not sure about the name, maybe something like: `MAX_UNEVALUATED_DF_ROWS`. And maybe add a short comment saying something like: `Maximum number of rows to request from an unevaluated (out-of-core) dataframe`

sfc-gh-tszerszen replied:
It's updated 👍

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -440,6 +444,11 @@ def convert_anything_to_df(df: Any) -> DataFrame:
     if is_type(df, "numpy.ndarray") and len(df.shape) == 0:
         return pd.DataFrame([])
 
+    if is_type(df, "snowflake.snowpark.dataframe.DataFrame") and not isinstance(
```
nit: I think you can use the constant here: `_SNOWPARK_DF_TYPE_STR`

sfc-gh-tszerszen replied:
It's updated 👍

lukasmasuch commented on lib/streamlit/elements/arrow.py:
```diff
@@ -145,6 +154,21 @@ def dg(self) -> "DeltaGenerator":
         return cast("DeltaGenerator", self)
 
 
+def show_warning_if_data_is_capped(data: Data = None) -> Data:
```
I think we can move this to `type_utils` as well.

lukasmasuch replied:
Ahh btw, I think it might be sufficient to actually do the error message handling directly in the `convert_anything_to_df`. In that case, we don't have to call this in all the individual commands. Or was there an issue with executing `st.caption` in `convert_anything_to_df`?

lukasmasuch replied:
The thing is that `convert_anything_to_df` is anyways called in all our components having `dataframe` as input. So, the `caption` would be sent at the right time

sfc-gh-tszerszen replied:
I was worried that it will be called too many times, but seems like it's not the case 👍 Good call out, it's updated and it's much cleaner 👍

sfc-gh-tszerszen commented:
Ok, it's updated, can you please re-review it? 🙇 cc: @LukasMasuch

lukasmasuch commented on lib/streamlit/elements/arrow.py:
```diff
@@ -30,14 +30,23 @@
     from streamlit.delta_generator import DeltaGenerator
 
-Data = Union[DataFrame, Styler, pa.Table, ndarray, Iterable, Dict[str, List[Any]], None]
+Data = Union[
```
nit: I think this could be formatted back to the existing formatting since nothing was actually changed here

lukasmasuch commented on lib/streamlit/elements/arrow_altair.py:
```diff
@@ -127,6 +127,7 @@ def _arrow_line_chart(
            height: 220px
 
         """
+
```
nit: can you clean up the empty lines, it's probably just a relict from the previous changes

lukasmasuch reviewed (approved):
LGTM 👍 But can you add one more test for `convert_anything_to_df` to test this with the snowpark dataframe mock & also test that it actually get capped at `10k` rows if the mock has `11k` rows.

lukasmasuch commented on lib/streamlit/elements/dataframe_selector.py:
```diff
@@ -47,7 +47,7 @@ def dataframe(
         Parameters
         ----------
-        data : pandas.DataFrame, pandas.Styler, pyarrow.Table, numpy.ndarray, Iterable, dict, or None
+        data : pandas.DataFrame, pandas.Styler, pyarrow.Table, numpy.ndarray, snowflake.snowpark.DataFrame, Iterable, dict, or None
```
And `line_chart`, `bar_chart`, `area_chart`, `table` also need to get this docstring addition

sfc-gh-tszerszen replied:
It's updated now 👍

lukasmasuch commented on e2e/scripts/st_arrow_dataframe_unevaluated_snowpark_dataframe.py:
```diff
@@ -0,0 +1,18 @@
+import streamlit as st
+from tests.streamlit.snowpark_mocks import DataFrame
+
+st.dataframe(DataFrame())
```
nit: Probably better to actually use more rows than 10k in this mock test

sfc-gh-tszerszen replied:
Now it's using ```snowpark_dataframe = DataFrame(num_of_rows=50000)``` 👍

lukasmasuch commented on e2e/scripts/st_arrow_dataframe_unevaluated_snowpark_dataframe.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
```
And can you rename this file to `st_arrow_unevaluated_snowpark_dataframe.py` and also test this for `line_chart`, `bar_chart`, `area_chart` and `table`

sfc-gh-tszerszen replied:
It's renamed now and almost all the tests you mentioned are present, ⚠️ with the exception:

* st.table renders properly, however it takes so looooong to render
* The E2E tests for every PR already are taking so much time, I've decided to not test st.table, is that ok?

lukasmasuch replied:
yep, that's fine. Thought `st.table` might need some investigation on why it is so slow. I think it would speed it up if we use fewer columns (e.g. just 5)

lukasmasuch replied:
btw. what does slow mean, how long does it run (with & without `st.table`?)

sfc-gh-tszerszen replied:
Sure 👍 It's lowered to 4 now 👍

sfc-gh-tszerszen committed:
Snowpark integration: Add support for unevaluated dataframes

lukasmasuch commented on e2e/scripts/st_arrow_unevaluated_snowpark_dataframe.py:
```diff
@@ -0,0 +1,26 @@
+import streamlit as st
+from tests.streamlit.snowpark_mocks import DataFrame
+
+snowpark_dataframe = DataFrame(num_of_rows=50000)
```
You might need to set a seed for random:
`random.seed(0)`

Otherwise, you are always getting a different version of the dataframe.

sfc-gh-tszerszen replied:
It's updated 👍

sfc-gh-tszerszen committed:
use random.seed(0) and lower num of cols for the table

lukasmasuch commented on lib/streamlit/type_util.py:
```diff
@@ -440,6 +445,14 @@ def convert_anything_to_df(df: Any) -> DataFrame:
+        df = pd.DataFrame(df.take(MAX_UNEVALUATED_DF_ROWS))
+        if df.shape[0] == MAX_UNEVALUATED_DF_ROWS:
+            st.caption(
+                f"Showing only 10k rows. Call `collect()` on the dataframe to show more."
```
nit: @jrieke should there be a ⚠️ emoji or not?

lukasmasuch replied:
Just talked to @jrieke, and the ⚠️ should be added to the caption as a prefix.

sfc-gh-tszerszen replied:
@jrieke can confirm, but I'm adding it, because I think is should be there, quoting Johannes I assume the words:

> Change the warning message to

mean it should be clear that it's the warning, thus we need ⚠️, which is a good, informative user experience 

Thank you Lukas for mentioning it 👍

sfc-gh-tszerszen committed:
add the warning
